typed/remove_test.go

```diff
@@ -0,0 +1,340 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemoveDeduced(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{}`,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: ``,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			// this is okay, because in a deduced schema `b` is handled as a map not a struct
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": null}`,
+		},
+		{
+			object: `{"a": "value", "b": []}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			// TODO: this should fail
+			expect: `{"a": "value", "b": ["c"]}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c", "d"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": ["c", "d"]}`,
+		},
+		{
+			object: `{"a": "value", "b": [{"c": "value"}, {"d": "value"}]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": [{"c": "value"}, {"d": "value"}]}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value", "d":{"e":"value"}}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "d")),
+			expect: `{"a": "value", "b": {"c":"value"}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value", "d": {"e":"value"}}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "d", "e")),
+			expect: `{"a": "value", "b": {"c":"value", "d": null}}`,
```

Comment:
I would expect ...
```
expect: `{"a": "value", "b": {"c":"value", "d": {}}}`
```

Reply:
That's what we have to decide.
I think we had this discussion before.
What does it mean to null and what to {}.

Reply:
I would keep it as `{}` rather than null tbh. Because the goal is not to change the type or meaning of the parent object. Also we do typically give "null" a stronger meaning than empty list as far as I can tell, and I don't think that's good here, even though it proably doesn't make a huge difference.

Reply:
Okay, i'll change it and see what breaks.
Just for reference, i think that's one of the previous discussions about this: https://github.com/kubernetes-sigs/structured-merge-diff/pull/133#discussion_r364017082